parser/prism/Translator.cc

```diff
@@ -362,7 +362,74 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node, bool preserveCon
 
             auto sorbetElements = translateMulti(arrayNode->elements);
 
-            return make_unique<parser::Array>(location, move(sorbetElements));
+            if (!directlyDesugar || !hasExpr(sorbetElements)) {
+                return make_unique<parser::Array>(location, move(sorbetElements));
+            }
+
+            auto locZeroLen = location.copyWithZeroLength();
+
+            ast::Array::ENTRY_store elems;
+            elems.reserve(sorbetElements.size());
+            ExpressionPtr lastMerge;
+            for (auto &stat : sorbetElements) {
+                if (parser::NodeWithExpr::isa_node<parser::Splat>(stat.get()) ||
+                    parser::NodeWithExpr::isa_node<parser::ForwardedRestArg>(stat.get())) {
+                    // Desugar [a, *x, remaining] into a.concat(<splat>(x)).concat(remaining)
+
+                    // The Splat was already desugared to Send{Magic.splat(arg)} with the splat's own location.
+                    // But for array literals, we want the splat to have the array's location to match
+                    // the legacy parser's behavior (important for error messages and hover).
+                    auto splatExpr = stat->takeDesugaredExpr();
+
+                    // The parser::Send case makes a fake parser::Array with locZeroLen to hide callWithSplat
+                    // methods from hover. Using the array's loc means that we will get a zero-length loc for
+                    // the Splat in that case, and non-zero if there was a real Array literal.
+                    if (auto send = ast::cast_tree<ast::Send>(splatExpr)) {
+                        ENFORCE(send->numPosArgs() == 1, "Splat Send should have exactly 1 argument");
+                        // Extract the argument from the old Send and create a new one with array's location
+                        splatExpr = MK::Splat(location, move(send->getPosArg(0)));
+                    }
```

Comment:
I am happy to get this landed...but it seems like not creating an extraneous `Send` here would be better, and going with the original "modify the location of the prism node" that the `Desugar.cc` code did.  We can do that in a followup -- I was not being especially clear about what the preferred course here was.